lib/chef/provider/file.rb

```diff
@@ -67,7 +67,13 @@ class File < Chef::Provider
       def initialize(new_resource, run_context)
         @content_class ||= Chef::Provider::File::Content
         if new_resource.respond_to?(:atomic_update)
-          @deployment_strategy = Chef::FileContentManagement::Deploy.strategy(new_resource.atomic_update)
+          # Resolves https://github.com/chef/chef/issues/4949
+          if docker_guest?(run_context.node) && (new_resource.path == "/etc/resolv.conf" ||
+                                                 new_resource.path == "/etc/hostname")
+            @deployment_strategy = Chef::FileContentManagement::Deploy.strategy(false)
+          else
+            @deployment_strategy = Chef::FileContentManagement::Deploy.strategy(new_resource.atomic_update)
```

Comment:
I feel like an explicit value in `new_resource.atomic_updates` should always override, just that the default becomes path-dependent.